### ISSUE TITLE
Add Lightning-like Profiler to Keras

### ISSUE THREAD

**Issue by innat** [keras-team/keras#17273]
**System information**.

TensorFlow version (you are using): 2.9
Are you willing to contribute it (Yes/No) : No 

## Describe the feature and the current behavior/state

In PyTorch-Lightning, you can identify bottlenecks in your code using [**Profiler**](https://pytorch-lightning.readthedocs.io/en/1.6.0/advanced/profiler.html) API. And I've found it super useful and easy to inspect the pros and cons of the code. This API offers 3 type of inferface, namely `simple`, `advance` and `pytorch`. I mostly use `simple` that summarize the execution time of medium to high level training + validation + inference API. The saved log files look something as follows. 

```python
# by passing a string
trainer = Trainer(..., profiler="simple")

# or by passing an instance
from pytorch_lightning.profiler import SimpleProfiler

profiler = SimpleProfiler()
trainer = Trainer(..., profiler=profiler)
```

![a](https://user-images.githubusercontent.com/17668390/204000492-0e7c8ee3-31ba-4194-bde1-acca200b2a23.png)

I don't know, in `keras`, if we have anything like this already. By having such high-level functionality in `keras` would really great. It will help to inspect the bottleneck of dataloader, model, callbacks etc, by reporting the time and memory consumption in the system.

I looked for it before, [discussion](https://discuss.tensorflow.org/t/profiler-report-like-pytorch-lighting-in-tf-keras-interface/3768). 
 
## Will this change the current api? How?

This can be considered as a [Callback](https://keras.io/api/callbacks/), I think. 

```python
tf.keras.callbacks.Profiler
or
tf.keras.utils.Profiler
```

**Who will benefit from this feature?**

The keras community. Engineers and researchers, both. Mostly engineers in production phase for code optimization.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Comment by innat** [comment]
cc. @bhack

**Comment by bhack** [comment]
We had a long thread at https://github.com/keras-team/keras/issues/16699

Also using the standard Tensorboard it is hard to precisely profile single `tf.function`:
https://github.com/tensorflow/profiler/issues/503

**Comment by bhack** [comment]
What is your proposal? Do you want to have an alternative UI for the profiler portobufs?

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tsl/profiler/protobuf

**Comment by innat** [comment]
@bhack 
Sorry, I don't have any proposal to present here. The thing I can request to have is from UX, described [above](https://github.com/keras-team/keras/issues/17273#issuecomment-1333455350). Looking back to [here](https://github.com/keras-team/keras/issues/17273#issuecomment-1327529669), it feels a complex issue in general. I am also interested to know what @rchao @fchollet thoughts about such possible feature.

**Comment by innat** [comment]
No, haven't seen. Is it generated from tf profiler? Or, custom [from](https://github.com/keras-team/keras-cv/blob/master/benchmarks/vectorization_strategy_benchmark.py)?

**Comment by bhack** [comment]
> No, haven't seen. Is it generated from tf profiler? Or, custom [from](https://github.com/keras-team/keras-cv/blob/master/benchmarks/vectorization_strategy_benchmark.py)?

https://github.com/keras-team/keras-cv/issues/17#issuecomment-1021962006

**Comment by innat** [comment]
cc. @rchao 
A gentle reminder for this request.

**Comment by rchao** [comment]
This appears to be useful to me, but it's unclear whether there is available bandwidth on the team since it's some substantial work. Happy to check with the team and circle back in a couple of weeks.

**Comment by bhack** [comment]
@rchao Do you know something about the mentioned internal tool?

**Comment by rchao** [comment]
Correct, there has been experimental efforts on a profiling callback that hasn't been open sourced yet. That's one thing I'll also check with the team.
